DPKI bootstrap (#1182)
holochain/holochain-rust: doc/holochain_101/src/dpki.md

# Distributed Public Key Infrastructure (DPKI)

## Context

> Like most other distributed systems, Holochain fundamentally relies on public key cryptography. Among other uses, Holochain nodes are identified by their public keys, and thus provenance of messages from nodes can be checked simply by checking a message signature against the node's identifier.  This fundamental use requires nodes to keep the private key secret, lest the node's very agency be compromised.  Keeping such secrets in the digital age is a non-trivial problem.  Additionally, distributed systems create a context of a profusion of nodes with multiple public keys, many of which may wish to be identified as being under the control a of a single actor.  Solutions to this need create yet another layer of keys which sign other keys, and the management of all this can become quite complex.

Reviewer: ```suggestion
Like most other distributed systems, Holochain fundamentally relies on public key cryptography. Among other uses, Holochain nodes are identified by their public keys, and thus provenance of messages from nodes can be checked simply by checking a message signature against the node's identifier.  This fundamental use requires nodes to keep the private key secret, lest the node's very agency be compromised.  Keeping such secrets in the digital age is a non-trivial problem.  Additionally, distributed systems imply an abundance of nodes with public keys, many of which may wish to be identified as being under the control of a single actor.  Solutions to this need create yet another layer of keys which sign other keys, and the management of all this can become quite complex.
```

## DeepKey

To deliver on the basics of Distributed Public Key Infrastructure, we need a way to generate keys of various types (revocation, identity, encryption, signing) from seeds, and we need to be able to generate such seeds from primary seeds, so that a human agent can create related "device agents" provably under their control.

After studying a number of uses cases, including initial sign-up, key revocation, etc, the central insight we came to was the need to create a Hierarchical Deterministic Key generation system, based on a Primary Seed, from which additional seeds can be generated which then are in turn used to actually generate many key-pairs.  This allows us, by-convention, to use the first seed generated by the Primary seed as the seed for revocation keys, and subsequent seeds as seeds for keys of separate Holochain devices that can be proven to be under the control of the holder of Primary Seed.

DeepKey

TODO: merge the various docs we developed to explain DeepKey here.
- https://medium.com/holochain/part-2-holochain-holo-accounts-cryptographic-key-management-and-deepkey-bf32ee91af65

> - https://hackmd.io/UbfvwQdJRKaAHI9Xa7F3VA?view

Reviewer: Are these hackmds editable by anyone who comes upon them?  We should check the permissions, since this will be published

- https://hackmd.io/8c8rZCyaTTqH_7TIBVtEUQ
- https://hackmd.io/oobu0sKMSMadLXza4rHY_g

## DPKI Implementation Technical Details

### Keystore
For each Holochain DNA instance, the Conductor maintains a Keystore, which holds "secrets" (seeds and keys) needed for cryptographic signing and encrypting. Each of the secrets in the Keystore is associated with a string which is a handle needed when using that secret for some cryptographic operation.  Our cryptographic implementation is based on libsodium, and the seeds use their notions of context and index for key derivation paths.  This implementation allows DNA developers to securely call cryptographic functions from WASM which will be executed in the conductor's secure memory space when actually doing the cryptographic processing.

> Decrypting a secret from the keystore invokes a passphrase manager service, which is used to collect the passphrase from some end-user.  This service is implementation specific.  Currently we have implementations for command-line passphrase collection for use in the `hc keyget` command, and also an implementation that routes through the same JSON RPC pathway for other admin commands sent to the conductor for other admin actions.

Reviewer 2: We don't have an implementation for passphrase through JSON RPC yet...
Also, this should it should be `keygen` instead of `keyget`, right?
```suggestion
Decrypting a secret from the keystore invokes a passphrase manager service, which is used to collect the passphrase from some end-user.  This service is implementation specific.  Currently we have implementations for command-line passphrase collection for use in the `hc keygen` command, and also a command-line implementation within the conductor.
```
